tests/system/utils.py

```diff
@@ -143,3 +143,75 @@ def convert_pandas_dtypes(df: pd.DataFrame, bytes_col: bool):
     df["numeric_col"] = df["numeric_col"].apply(
         lambda value: decimal.Decimal(str(value)) if value else None  # type: ignore
     )
+
+
+def assert_pandas_df_equal_pca_components(actual, expected, **kwargs):
+    """Compare two pandas dataframes representing PCA components. The columns
+    required to be present in the dataframes are:
+        numerical_value: numeric,
+        categorical_value: List[object(category, value)]
+
+    Args:
+        actual: Actual Pandas DataFrame
+
+        expected: Expected Pandas DataFrame
+
+        kwargs: kwargs to use in `pandas.testing.assert_series_equal` per column
+    """
+    # Compare the index, columns and values separately, as the polarity of the
+    # PCA vectors can be arbitrary
+    pd.testing.assert_index_equal(
+        actual.index, expected.index.astype(actual.index.dtype)
+    )  # dtype agnostic index comparison
+    pd.testing.assert_index_equal(actual.columns, expected.columns)
+    for column in expected.columns:
+        try:
+            pd.testing.assert_series_equal(actual[column], expected[column], **kwargs)
+        except AssertionError:
+            if column not in {"numerical_value", "categorical_value"}:
+                raise
+
+            # Allow for sign difference per numeric/categorical column
+            if column == "numerical_value":
```

Comment:
I guess the reason we compare every column is because non-numerical columns can't apply abs() operation. 

1, For numerical columns, it is still neater to call abs() then compare than try compare and catch failures.
2, IMO it is more readable to construct the normalized DFs first then just compare them equal. Rather than write a customized compare method. Also it is easier to debug, we can directly compare the normalized DFs and see the diff.

Reply:
The main reason not doing `abs` is that the valid PCs can differ in polarity of the whole vector, e.g. 

this is acceptable: [1, -2, 3] vs [-1, 2, -3]

but this is not: [1, -2, 3] vs [1, 2, 3]

`abs` will pass even the latter.

Reply:
It won't be a problem, every axis is perpendicular and independent to each other. The later is a valid output too.

Reply:
As discussed offline, below is a real example of a failing test. the second column is exactly the same, while the first and the third columns differ in polarity. With `abs` we will relax more than we need to, which would pass cases even if only some (not all) columns differ in polarity, which we don't want.

---------------------------

expected
            principal_component_1  principal_component_2  principal_component_3
tag_number                                                                     
1633                    -1.314041              -0.889106              -0.704345
1672                    -0.855813              -1.259753               0.322555
1690                    -1.848786              -0.983304              -0.095759
actual
            principal_component_1  principal_component_2  principal_component_3
tag_number                                                                     
1633                     1.314041              -0.889106               0.704345
1672                     0.855813              -1.259753              -0.322555
1690                     1.848786              -0.983304               0.095759